File: python/problems/leetcode/0024_swap_nodes_in_pairs/solution.py
"""LeetCode 24: Swap Nodes in Pairs."""

from __future__ import annotations

from typing import Optional


class ListNode:
    def __init__(self, val: int = 0, next: Optional["ListNode"] = None) -> None:
        self.val = val
        self.next = next


class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        pre = dummy
        while pre.next is not None and pre.next.next is not None:
            a = pre.next
            b = a.next
            a.next = b.next
            b.next = a
            pre.next = b
            pre = a
        return dummy.next


def from_list(vs: list[int]) -> Optional[ListNode]:
    dummy = ListNode()
    cur = dummy
    for v in vs:
        cur.next = ListNode(v)
        cur = cur.next
    return dummy.next


def to_list(h: Optional[ListNode]) -> list[int]:
    out: list[int] = []
    while h is not None:
        out.append(h.val)
        h = h.next
    return out


if __name__ == "__main__":
    assert to_list(Solution().swapPairs(from_list([1, 2, 3, 4]))) == [2, 1, 4, 3]
    assert to_list(Solution().swapPairs(from_list([1]))) == [1]
    assert Solution().swapPairs(None) is None
    print("0024 OK")

File: cpp/problems/leetcode/0024_swap_nodes_in_pairs/solution.cpp
// LeetCode 24
#include <alg_std.hpp>
#include <cassert>
using namespace std;

struct ListNode {
    int val;
    ListNode* next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int v) : val(v), next(nullptr) {}
    ListNode(int v, ListNode* n) : val(v), next(n) {}
};

class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        ListNode dummy(0, head);
        ListNode* pre = &dummy;
        while (pre->next && pre->next->next) {
            ListNode* a = pre->next;
            ListNode* b = a->next;
            a->next = b->next;
            b->next = a;
            pre->next = b;
            pre = a;
        }
        return dummy.next;
    }
};

static ListNode* fromList(vector<int> vs) {
    ListNode dummy;
    ListNode* cur = &dummy;
    for (int v : vs) {
        cur->next = new ListNode(v);
        cur = cur->next;
    }
    return dummy.next;
}

static vector<int> toList(ListNode* h) {
    vector<int> out;
    while (h) {
        out.push_back(h->val);
        h = h->next;
    }
    return out;
}

int main() {
    assert((toList(Solution().swapPairs(fromList({1, 2, 3, 4}))) ==
            vector<int>{2, 1, 4, 3}));
    assert((toList(Solution().swapPairs(fromList({1}))) == vector<int>{1}));
    assert(Solution().swapPairs(nullptr) == nullptr);
    cout << "0024 OK" << endl;
    return 0;
}
